Data/HashMap/Base.hs

```diff
@@ -345,13 +354,15 @@ cmp cmpk cmpv t1 t2 = go (toList' t1 []) (toList' t2 [])
       = compare k1 k2 `mappend`
         leafCompare l1 l2 `mappend`
         go tl1 tl2
-    go (Collision k1 ary1 : tl1) (Collision k2 ary2 : tl2)
-      = compare k1 k2 `mappend`
-        compare (A.length ary1) (A.length ary2) `mappend`
-        unorderedCompare leafCompare (A.toList ary1) (A.toList ary2) `mappend`
+    go (Collision h1 (L k11 v11) (L k12 v12) hm1 : tl1) (Collision h2 (L k21 v21) (L k22 v22) hm2 : tl2)
+      = compare h1 h2 `mappend`
+        compare (length hm1) (length hm2) `mappend`
+        unorderedCompare (\(k1, v1) (k2, v2) -> (k1 `cmpk` k2) `mappend` (v1 `cmpv` v2))
+          ((k11, v11) : (k12, v12) : toList hm1)
+          ((k21, v21) : (k22, v22) : toList hm2) `mappend`
```

Comment:
I'm not at all convinced that `unorderedCompare` is sure to produce a total ordering from two total orderings at all. It looks completely bonkers to me. Unless you can give me a good argument for this function making any sense whatsoever, I think we should *delete* the `Ord2` instance altogether and write a reasonable implementation of `Ord1`.

Reply:
I've asked Oleg if he can explain his reasoning. If it proves solid, we'll have to deal with it, but note in the documentation that `liftCompare2` can have off-the-charts bad performance in pathological cases. Regardless, we should surely consider a simpler and more efficient implementation of the inherently better-behaved `liftCompare`.

Reply:
@phadej has half an explanation and lots of QuickCheck testing. So maybe we can back-burner the general validity question for now. I'm still deeply uncomfortable about the fact that I don't know exactly what relationship must exist among `==`, `hashWithSalt`, and the passed comparison function to make `liftCompare2` behave reasonably.

Reply:
@treeowl Looking at the diff between the old code and the new code, I don't think the new code is asymptotically less performant. What's the next action here?

Reply:
You misunderstand. If someone is actually using this instance, and encounters a map with a huge collision bucket, then they will face a problem like the one you're trying to address. I don't think we can fix that for `Ord2` (even though the instance appears to be moderately validish). But we most certainly *can* fix it for `Ord1` and `Ord`, so we should separate those implementations.